File: backend/app/tree_grid.py
from __future__ import annotations

import json
import math
from functools import lru_cache
from pathlib import Path

from .models import LatLngLiteral, TreeGridCell, TreeGridData

TREE_GRID_PATH = Path(__file__).resolve().parents[2] / "data" / "tree-grid.sample.json"


@lru_cache(maxsize=1)
def get_tree_grid() -> TreeGridData:
    with TREE_GRID_PATH.open("r", encoding="utf-8") as handle:
        payload = json.load(handle)
    return TreeGridData.model_validate(payload)


@lru_cache(maxsize=1)
def build_cell_lookup():
    return {cell.key: cell for cell in get_tree_grid().cells}


def get_grid_key(point: LatLngLiteral, grid: TreeGridData | None = None):
    selected_grid = grid or get_tree_grid()
    lat_index = math.floor((point.lat - selected_grid.origin.lat) / selected_grid.latStep)
    lng_index = math.floor((point.lng - selected_grid.origin.lng) / selected_grid.lngStep)
    return f"{lat_index}:{lng_index}"


def lookup_tree_cell(point: LatLngLiteral, grid: TreeGridData | None = None) -> TreeGridCell | None:
    selected_grid = grid or get_tree_grid()
    return build_cell_lookup().get(get_grid_key(point, selected_grid))

File: lib/tree-grid.ts
import treeGridData from "@/data/tree-grid.sample.json";
import type { LatLngLiteral, TreeGridCell, TreeGridData } from "@/lib/types";
import { distanceMeters } from "@/lib/utils";

let cachedLookup: Map<string, TreeGridCell> | null = null;

export function getTreeGrid(): TreeGridData {
  return treeGridData as unknown as TreeGridData;
}

export function buildCellLookup(grid = getTreeGrid()): Map<string, TreeGridCell> {
  if (!cachedLookup) {
    cachedLookup = new Map(grid.cells.map((cell) => [cell.key, cell]));
  }

  return cachedLookup;
}

export function getGridKey(point: LatLngLiteral, grid = getTreeGrid()): string {
  const latIndex = Math.floor((point.lat - grid.origin.lat) / grid.latStep);
  const lngIndex = Math.floor((point.lng - grid.origin.lng) / grid.lngStep);
  return `${latIndex}:${lngIndex}`;
}

export function lookupTreeCell(point: LatLngLiteral, grid = getTreeGrid()): TreeGridCell | null {
  return buildCellLookup(grid).get(getGridKey(point, grid)) ?? null;
}

/**
 * Returns all cells whose centers fall within `radiusMeters` of `point`.
 * With a fine-resolution grid (e.g. ~20 m cells) this captures every tree
 * that a pedestrian at that point could realistically be exposed to.
 */
export function lookupTreeCellsInRadius(
  point: LatLngLiteral,
  radiusMeters: number,
  grid = getTreeGrid(),
): TreeGridCell[] {
  const lookup = buildCellLookup(grid);
  const latIndex = Math.floor((point.lat - grid.origin.lat) / grid.latStep);
  const lngIndex = Math.floor((point.lng - grid.origin.lng) / grid.lngStep);

  // How many cell steps to scan in each axis (~111 km/deg lat, ~84 km/deg lng at NYC)
  const latRange = Math.ceil(radiusMeters / (grid.latStep * 111_000)) + 1;
  const lngRange = Math.ceil(radiusMeters / (grid.lngStep * 84_000)) + 1;

  // The primary cell (the one that contains the point) is always included —
  // the pedestrian is literally inside it, regardless of distance to its center.
  const primaryKey = `${latIndex}:${lngIndex}`;
  const results: TreeGridCell[] = [];

  for (let dlat = -latRange; dlat <= latRange; dlat++) {
    for (let dlng = -lngRange; dlng <= lngRange; dlng++) {
      const key = `${latIndex + dlat}:${lngIndex + dlng}`;
      const cell = lookup.get(key);
      if (!cell) continue;
      if (key === primaryKey || distanceMeters(point, cell.center) <= radiusMeters) {
        results.push(cell);
      }
    }
  }

  return results;
}
